examples/win-proc-stat.c

```diff
@@ -0,0 +1,476 @@
+{
+    if (!event->data) {
+        printf("Null pointer to context!\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    GHashTable* stats = ((struct cb_context*)event->data)->stats;
+
+    vmi_pid_t pid = -1;
+    vmi_dtb_to_pid(vmi, event->reg_event.value, &pid);
+    int32_t tid = get_current_thread_id(vmi, event);
+
+    if (pid >= 0 && tid >= 0) {
+        pid_tid_t key = { .pid = pid, .tid = tid };
+        uint64_t value = (uint64_t)g_hash_table_lookup(stats, (gpointer)key.value) + 1;
+        g_hash_table_insert(stats, (gpointer)key.value, (gpointer)value);
+    } else
+        printf("Failed to get PID (%i) or TID (%i)\n", pid, tid);
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static volatile int interrupted = 0;
+static void close_handler(int sig)
+{
+    interrupted = sig;
+}
+
+int main(int argc, char** argv)
+{
+    /* this is the VM that we are looking at */
+    if (argc != 5) {
+        printf("Usage: %s name|domid <domain name|domain id> -r <rekall profile>\n", argv[0]);
+        return 1;
+    }
+
+    void* domain;
+    uint64_t domid = VMI_INVALID_DOMID;
+    uint64_t init_flags =  VMI_INIT_EVENTS;
+
+    if (strcmp(argv[1], "name")==0) {
+        domain = (void*)argv[2];
+        init_flags |= VMI_INIT_DOMAINNAME;
+    } else if (strcmp(argv[1], "domid")==0) {
+        domid = strtoull(argv[2], NULL, 0);
+        domain = (void*)&domid;
+        init_flags |= VMI_INIT_DOMAINID;
+    } else {
+        printf("You have to specify either name or domid!\n");
+        return 1;
+    }
+
+    char* rekall_profile = NULL;
+
+    if (strcmp(argv[3], "-r") == 0) {
+        rekall_profile = argv[4];
+    } else {
+        printf("You have to specify path to rekall profile!\n");
+        return 1;
+    }
+
+    /* for a clean exit */
+    struct sigaction act;
+
+    act.sa_handler = close_handler;
+    act.sa_flags = 0;
+    sigemptyset(&act.sa_mask);
+    sigaction(SIGHUP,  &act, NULL);
+    sigaction(SIGTERM, &act, NULL);
+    sigaction(SIGINT,  &act, NULL);
+    sigaction(SIGALRM, &act, NULL);
+
+    /* Some local variables */
+    struct cb_context* ctx = NULL;
+    GHashTable* stats = NULL;
+
+    /* Get offsets */
+    addr_t* offsets = fill_offsets_from_rekall(rekall_profile);
+    if (!offsets)
+        return 1;
+
+    /* Prepare config */
+    GHashTable* config = NULL;
+    config = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+    if (!config) {
+        printf("Failed to create GHashTable!\n");
+        goto done;
+    }
+    g_hash_table_insert(config, g_strdup("os_type"), g_strdup("Windows"));
+    g_hash_table_insert(config, g_strdup("rekall_profile"), g_strdup(rekall_profile));
+
+    /* initialize the libvmi library */
+    vmi_instance_t vmi = NULL;
+    vmi_init_error_t err = VMI_SUCCESS;
+
+    if (VMI_FAILURE ==
+            vmi_init_complete(&vmi, domain, init_flags, NULL,
+                              VMI_CONFIG_GHASHTABLE, config, &err)) {
```

Comment:
You can pass in the path directly to `vmi_init_complete` now, no need to create the GHashTable here. Just use VMI_CONFIG_JSON_PATH as the identifier. LibVMI can figure out the OS type from the profile too.

Reply:
@tklengyel, done!